File: practice/algorithms/implementation/utopian_tree/utopian_tree.rb
def tree(actualCycle, cycles, height)
  return height if(actualCycle == cycles)
  return tree(actualCycle + 1, cycles, height * 2) if(actualCycle % 2 == 0)
  tree(actualCycle + 1, cycles, height + 1)
end

tests = gets.to_i
tests.times do
  cycles = gets.to_i
  puts tree(0, cycles, 1)
end

File: practice/algorithms/implementation/utopian_tree/UtopianTree.js
function processData(input) {
    var numbers = input.split("\n").map(function(i) {return parseInt(i); });
    var index = 0;
    var tests = numbers[index++];

    for(var i = 0; i < tests; i++) {
        var cycles = numbers[index++];
        console.log(tree(0, cycles, 1));
    }
}

function tree(actual, cycles, height) {
    if(actual == cycles) {
        return height
    }
    if(actual % 2 == 0)
        return tree(actual + 1, cycles, height * 2)
    return tree(actual + 1, cycles, height + 1)
}

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
    processData(_input);
});